odl/space/cupy_tensors.py

```diff
@@ -0,0 +1,2372 @@
+            Default: `RealNumbers`
+
+        Returns
+        -------
+        dtype : `numpy.dtype`
+            Numpy data type specifier. The returned defaults are:
+
+            - ``RealNumbers()`` or ``None`` : ``np.dtype('float64')``
+            - ``ComplexNumbers()`` : ``np.dtype('complex128')``
+
+            These choices correspond to the defaults of the ``cupy``
+            library.
+        """
+        if field is None or field == RealNumbers():
+            return np.dtype('float64')
+        elif field == ComplexNumbers():
+            return np.dtype('complex128')
+        else:
+            raise ValueError('no default data type defined for field {}.'
+                             ''.format(field))
+
+
+class CupyTensor(Tensor):
+
+    """Representation of an `CupyTensorSpace` element."""
+
+    def __init__(self, space, data):
+        """Initialize a new instance."""
+        super(CupyTensor, self).__init__(space)
+        self.__data = data
+
+    @property
+    def data(self):
+        """Raw `cupy.core.core.ndarray` representing the data."""
+        return self.__data
+
+    @property
+    def ndim(self):
+        """Number of axes (=dimensions) of this tensor."""
+        return self.space.ndim
+
+    @property
+    def device(self):
+        """The GPU device on which this tensor lies."""
+        return self.space.device
+
+    def asarray(self, out=None):
+        """Extract the data of this element as a `numpy.ndarray`.
+
+        Parameters
+        ----------
+        out : `numpy.ndarray`, optional
+            Array to which the result should be written.
+            Has to be contiguous and of the correct data type.
+
+        Returns
+        -------
+        asarray : `numpy.ndarray`
+            Numpy array of the same `dtype` and `shape` this tensor.
+            If ``out`` was given, the returned object is a reference to it.
+
+        Examples
+        --------
+        By default, a new array is created:
+
+        >>> r3 = odl.rn(3, impl='cupy')
+        >>> x = r3.element([1, 2, 3])
+        >>> x.asarray()
+        array([ 1.,  2.,  3.])
+        >>> int_spc = odl.tensor_space(3, impl='cupy', dtype=int)
+        >>> x = int_spc.element([1, 2, 3])
+        >>> x.asarray()
+        array([1, 2, 3])
+        >>> tensors = odl.rn((2, 3), impl='cupy', dtype='float32')
+        >>> x = tensors.element([[1, 2, 3],
+        ...                      [4, 5, 6]])
+        >>> x.asarray()
+        array([[ 1.,  2.,  3.],
+               [ 4.,  5.,  6.]], dtype=float32)
+
+        Using the out parameter, the array can be filled in-place:
+
+        >>> out = np.empty((2, 3), dtype='float32')
+        >>> result = x.asarray(out=out)
+        >>> out
+        array([[ 1.,  2.,  3.],
+               [ 4.,  5.,  6.]], dtype=float32)
+        >>> result is out
+        True
+        """
+        if out is None:
+            return cupy.asnumpy(self.data)
+        else:
+            if out.shape != self.shape:
```

Comment:
If out is not a numpy array, this line will fail in weird manners

Reply:
It will say `[...] has no attribute 'shape'`, is that so weird?

Reply:
Well I guess that's OK actually, but what about other stuff (i.e. if you feed this a `NumpyTensor`?)

Reply:
Well yeah, I give up 😬 I'll do the instance check. After all, the docstring says a Numpy array is required.